sound/soc/intel/boards/sof_sdw.c

```diff
@@ -411,25 +411,30 @@ static int create_codec_dai_name(struct device *dev,
 
 static int set_codec_init_func(const struct snd_soc_acpi_link_adr *link,
 			       struct snd_soc_dai_link *dai_links,
-			       bool playback)
+			       bool playback, int group_id)
 {
 	int i;
 
-	for (i = 0; i < link->num_adr; i++) {
-		unsigned int part_id;
-		int codec_index;
-
-		part_id = SDW_PART_ID(link->adr_d[i].adr);
-		codec_index = find_codec_info_part(part_id);
+	do {
+		for (i = 0; i < link->num_adr; i++) {
+			unsigned int part_id;
+			int codec_index;
 
-		if (codec_index < 0)
-			return codec_index;
+			part_id = SDW_PART_ID(link->adr_d[i].adr);
+			codec_index = find_codec_info_part(part_id);
 
-		if (codec_info_list[codec_index].init)
-			codec_info_list[codec_index].init(link, dai_links,
-						 &codec_info_list[codec_index],
-						 playback);
-	}
+			if (codec_index < 0)
+				return codec_index;
+			if (link->adr_d[i].endpoints->group_id != group_id)
```

Comment:
@bardliao dont you have to check if the enpoint is aggregated here as well?

Reply:
> @bardliao dont you have to check if the enpoint is aggregated here as well?

group_id starts with 1 when it is aggregated. In other words, the aggregated flag is always 0 when group_id is 0. Also, ``while (link->mask && group_id)`` will be invalid if a codec is aggregated with group_id 0. In my opinion, we should fix it in the match table if it sets an non zero group id to a non aggregated endpoint or a zero group id to an aggregated endpoint.

Reply:
@bardliao but you have a do..while{} loop, so the first iteration could be even with group_id == 0 no?

Reply:
> @bardliao but you have a do..while{} loop, so the first iteration could be even with group_id == 0 no?

Yes, we should do codec init in the first iteration no matter it is aggregated or not.

Reply:
ahh now I get it. The commit message confused me a bit. I think you should just say: "Init codecs which belong to the same group id on all links".

Reply:
> ahh now I get it. The commit message confused me a bit. I think you should just say: "Init codecs which belong to the same group id on all links".

Done.